### PR TITLE
Add a katello_host_collection module

### PR DESCRIPTION
**Pull request by ephracis**
Add a new module for managing host collections. The module
katello_activation_key is also modified to support associating keys with
host collections.

This replaces #225

### CODE REVIEW

**Comment by ephracis** [comment]
@mdellweg Let me know if I'm close. I didn't manage to apply the diff so I had to manually copy-and-paste everything.

If there's anything missing let me know and I'll fix it.

File: modules/katello_host_collection.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'name',
+    'organization': 'organization',
+    'description': 'description',
+}
+
+
+def main():
+    module = AnsibleModule(
```
Comment (mdellweg) [inline modules/katello_host_collection.py]: in module_utils/foreman_helper, there is a `KatelloEntityAnsibleModule` to inherit from. Its use removes a lot of boilerplate here.
Also, if you feel adventurous, we started very recently to migrate the modules to a new api layer. In that case you would inherit from `KatelloEntityApipieAnsibleModule`. That is not yet required.

File: host_collection-0.yml [hunk 2]
```diff
@@ -0,0 +1,240 @@
+interactions:
```
Comment (mdellweg) [inline host_collection-0.yml]: Those files are definitely in the wrong directory. Anyway, they should be automatically generated by `make record_host_collection` in which case, they end up in the right place.

File: host_collection.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+---
```
Comment (mdellweg) [inline host_collection.yml]: I guess, this belongs into `test/test_playbooks`.

File: host_collection.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      host_collection_state: present
+      expected_change: true
+  - include: tasks/host_collection.yml
+    vars:
+      host_collection_state: present
+      expected_change: false
+  - include: tasks/host_collection.yml
+    vars:
+      host_collection_state: absent
+      expected_change: true
```
Comment (mdellweg) [inline host_collection.yml]: Can you repeat the absent task with `expected_change: false`, please. We test those modules for idempotence that way.

File: test/test_playbooks/tasks/activation_key_copy.yml [hunk 5]
```diff
@@ -21,7 +22,7 @@
     content_view: "{{  activation_key_content_view }}"
     subscriptions: "{{ activation_key_subscriptions }}"
     auto_attach: "{{ activation_key_auto_attach }}"
-    state: 'copied'
+    state: "{{ activation_key_state }}"
```
Comment (mdellweg) [inline test/test_playbooks/tasks/activation_key_copy.yml]: This file seems unneeded.

File: host_collection.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    vars:
+      organization_state: present
+
+- hosts: tests
+  gather_facts: false
+  tasks:
+  - name: Load server config
+    include_vars:
+      file: server_vars.yml
+  - include: tasks/host_collection.yml
```
Comment (mdellweg) [inline host_collection.yml]: I think, you are missing the `test/test_playbooks/tasks/host_collection.yml` file.
It should create a host collection with the same name (as the default value) as you are using in the activation key playbook.
Once you have all those files, you need to record the tests for host_collection and activation_key.
See: https://github.com/theforeman/foreman-ansible-modules/blob/master/README.md#how-to-test-modules-in-this-repository

File: modules/katello_host_collection.py (hunk 1 above)
Comment (ephracis) [reply modules/katello_host_collection.py]: Yeah I saw a few commits recently doing that. Let's move to that!

I'll change to inherit from `KatelloEntityApipieAnsibleModule` then. Anything I need to do when doing this?

Comment (mdellweg) [reply modules/katello_host_collection.py]: katello_content_credential is ported already. Take a look at that. It should in general become easier.

File: host_collection-0.yml (hunk 2 above)
Comment (ephracis) [reply host_collection-0.yml]: Haha, they sure are. The `patch` command didn't work quite well for me. Sorry I missed that. :)

File: test/test_playbooks/tasks/activation_key_copy.yml (hunk 5 above)
Comment (ephracis) [reply test/test_playbooks/tasks/activation_key_copy.yml]: Yeah I got that from the diff posted to #225. Should I revert that change and leave it as it is?

File: host_collection-0.yml (hunk 2 above)
Comment (mdellweg) [reply host_collection-0.yml]: just don't bother copy them. Drop em here, they are generated later.

File: host_collection.yml (hunk 6 above)
Comment (ephracis) [reply host_collection.yml]: Yup, I'm moving all the test files to their correct folder. Fixtures, playbooks and the missing task file.

File: host_collection-0.yml (hunk 2 above)
Comment (ephracis) [reply host_collection-0.yml]: Ok, so I should remove all 3 fixture files from the commit? Just add the playbook and task file under test then?

File: test/test_playbooks/tasks/activation_key_copy.yml (hunk 5 above)
Comment (mdellweg) [reply test/test_playbooks/tasks/activation_key_copy.yml]: Oh, i was too fast. That file _is_ needed. But i am not quite sure how it is meant to be used. In doubt leave it as is (not part of your pr...).

File: modules/katello_host_collection.py (hunk 1 above)
Comment (ephracis) [reply modules/katello_host_collection.py]: Ok, I've inherited from Apypie now instead. Let me know if I managed to get the boilerplate right.

File: modules/katello_host_collection.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            description=dict(),
+        ),
+        supports_check_mode=True,
+    )
+
+    (entity_dict, state) = module.parse_params()
+
+    module.connect()
+
+    entity_dict['organization'] = module.find_resource_by_name('organizations', entity_dict['organization'], thin=True)
```
Comment (mdellweg) [inline modules/katello_host_collection.py]: I believe, searching the organization is already done by `Katello.*Module`.

Comment (ephracis) [reply modules/katello_host_collection.py]: Ok, removing that line then.

Comment (mdellweg) [reply modules/katello_host_collection.py]: Hmm, not as such. But it is added to the argument_spec for you:
https://github.com/theforeman/foreman-ansible-modules/blob/master/module_utils/foreman_helper.py#L292

Comment (ephracis) [reply modules/katello_host_collection.py]: Oh, I should remove the line in the `argument_spec` then.

**Comment by ephracis** [comment]
Any idea why I am getting this in CI?

```
No match for the request (<Request (GET)
https://foreman.example.com/katello/api/v2/host_collections>) was found.
Can't overwrite existing cassette ('fixtures/activation_key-0.yml') in your current 
record mode ('none').
```

File: host_collection-0.yml (hunk 2 above)
Comment (mdellweg) [reply host_collection-0.yml]: Yes, drop them. They wont work with your new module anyway.
Before this pr is finished, you need to record those fixtures against a real Katello instance. Have a look at the README's How to test section.

**Comment by mdellweg** [comment]
Yes, you changed the activation_key module, thereby invalidating the recorded fixtures. You need to rerecord them. If you have the test setup (as specified in the README) up an running, it is a simple `make record_activation_key`. After that you need to check the newly generated fixtures into git.

**Comment by ephracis** [comment]
When I run `make record_activation_key` it seems to use `/usr/bin/python` but I have no pip for that one. Is there a way to force it to use `/usr/local/bin/python3` instead?

It complains I have no module yaml, but I do for Python 3.

**Comment by mdellweg** [comment]
I can only encourage you to use a python-venv for this.
Have a look at virtualenvwrapper

**Comment by ephracis** [comment]
Darn it. virtualenvwrapper wants to use `python` instead of `python3` as well. And it doesn't work with fish so I needed to revert back to bash. 😂

**Comment by mdellweg** [comment]
You can also use venv without the wrapper.
Anyway, good luck!

**Comment by ephracis** [comment]
Yeah I think I'm on track. Just managed to generate new activation key fixtures. I generated the host collection again as well, just in case.

**Comment by ephracis** [comment]
I ran `make test` and the upload test failed with `/bin/sh: rpm: command not found` which makes sense because I'm running on a Macbook right now.

But hopefully I'm a little bit closer now. Please take a look.

**Comment by ephracis** [comment]
Yay! It's all green now. Let me know if there's anything else I need to fix before this can be merged.

File: modules/katello_host_collection.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    state: present
+'''
+
+RETURN = ''' # '''
+
+try:
+    from ansible.module_utils.foreman_helper import (
+        sanitize_entity_dict,
+    )
+except ImportError:
```
Comment (mdellweg) [inline modules/katello_host_collection.py]: There is no use for this Exception catching anymore. Can you remove it, and add KatelloEntityApypieAnsibleModule here? (Only on import statement)

File: modules/katello_host_collection.py (hunk 7 above)
Comment (mdellweg) [reply modules/katello_host_collection.py]: :+1: exactly. I thouhgt i saw something, and it was the argument_spec.

File: test/test_playbooks/tasks/activation_key_copy.yml (hunk 5 above)
Comment (mdellweg) [reply test/test_playbooks/tasks/activation_key_copy.yml]: I meant, not adding it to your patch, and leaving the file as in master.

File: test/test_playbooks/activation_key.yml [hunk 9]
```diff
@@ -16,6 +16,9 @@
   - include: tasks/repository.yml
     vars:
       repository_state: present
+  - include: tasks/host_collection.yml
+    vars:
+      host_collection_state: present
```
Comment (mdellweg) [inline test/test_playbooks/activation_key.yml]: Can you please add some tests here?
immediately before deleting the activation_key, can you check that including the task with `activation_key_host_collections: omit` does not change anything (`expected_change`), while `activation_key_host_collections: []` deletes the associations (`expected_change: true`)?
Honestly, i am not quite sure, this functionality is already in the module, but i think, we want it.

File: modules/katello_host_collection.py (hunk 8 above)
Comment (ephracis) [reply modules/katello_host_collection.py]: You mean like this?

```python
from ansible.module_utils.foreman_helper import (
    sanitize_entity_dict,
    KatelloEntityApypieAnsibleModule
)
```

File: test/test_playbooks/activation_key.yml (hunk 9 above)
Comment (ephracis) [reply test/test_playbooks/activation_key.yml]: Absolutely. Adding this:

```yaml
  - include: tasks/activation_key.yml
    vars:
      activation_key_host_collections: omit
      expected_change: false
  - include: tasks/activation_key.yml
    vars:
      activation_key_host_collections: []
      expected_change: true
```

File: modules/katello_host_collection.py (hunk 8 above)
Comment (mdellweg) [reply modules/katello_host_collection.py]: Almost. I like to have a Comma even on the latest entry. It helps managing those lists line by line.

File: test/test_playbooks/activation_key.yml (hunk 9 above)
Comment (ephracis) [reply test/test_playbooks/activation_key.yml]: Sorry, not working.

> "No host collection found for omit"

Comment (ephracis) [reply test/test_playbooks/activation_key.yml]: Probably because it's hardcoded to be a list in the playbook. I'll see if I can fix it.

Comment (mdellweg) [reply test/test_playbooks/activation_key.yml]: Try with `...lection: "{{ omit }}"`

Comment (ephracis) [reply test/test_playbooks/activation_key.yml]: I think you were right. The `activation_key` module reports changed when host collection is omitted.

Comment (ephracis) [reply test/test_playbooks/activation_key.yml]:
```
TASK [debug] *******************************************************************
ok: [tests] => {
    "activation_key_host_collections": "__omit_place_holder__5f9bd43780e26a3d6e85302429661028a3df72b3"
}

TASK [Create/update katello activation key] ************************************
changed: [tests] => {
    "changed": true,
    "invocation": {
        "module_args": {
            "auto_attach": false,
            "content_overrides": [
                {
                    "label": "Test_Organization_Test_Product_Test_Repository",
                    "override": "disabled"
                }
            ],
            "content_view": "Default Organization View",
            "host_collections": null,
            "lifecycle_environment": "Library",
            "name": "Test Activation Key",
            "new_name": null,
            "organization": "Test Organization",
            "state": "present",
            "subscriptions": [
                {
                    "name": "Test Product"
                }
            ]
        }
    }
}

TASK [fail] ********************************************************************
task path: /Users/chbr/repos/work/github/foreman-ansible-modules/test/test_playbooks/tasks/activation_key.yml:33
fatal: [tests]: FAILED! => {
    "changed": false,
    "msg": "Creating/updating activation key failed! (expected_change: False)"
}
```

File: modules/katello_host_collection.py (hunk 8 above)
Comment (ephracis) [reply modules/katello_host_collection.py]: Done.

File: test/test_playbooks/activation_key.yml (hunk 9 above)
Comment (ephracis) [reply test/test_playbooks/activation_key.yml]: Ok I think I've managed to fix it. I removed it as a default and then explicitly added it into the test playbook for some of the tasks. I believe the playbook properly tests all cases now:
- Creating key
- Updating key without host collections
- Updating key with new (empty) host collections
- Deleting key

File: test/test_playbooks/tasks/activation_key.yml [hunk 10]
```diff
@@ -1,4 +1,5 @@
 ---
+- debug: var=activation_key_host_collections
```
Comment (mdellweg) [inline test/test_playbooks/tasks/activation_key.yml]: Please remove this debug statement again.

File: .gitignore [hunk 11]
```diff
@@ -7,3 +7,4 @@ __pycache__
 *.pyo
 *.swp
 .pytest_cache/
+venv/
```
Comment (mdellweg) [inline .gitignore]: I think, this safe.
A newline would be nice.

File: test/test_playbooks/tasks/activation_key_copy.yml [hunk 12]
```diff
@@ -2,6 +2,7 @@
 - name: "copy katello activation key"
   vars:
     activation_key_name: "Test Activation Key"
+    activation_key_state: "copied"
```
Comment (mdellweg) [inline test/test_playbooks/tasks/activation_key_copy.yml]: This is also not used.

File: test/test_playbooks/tasks/host_collection.yml [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    validate_certs: "{{ foreman_validate_certs }}"
+    name: "{{ host_collection_name }}"
+    description: "{{ host_collection_description }}"
+    organization: "{{ host_collection_organization }}"
+    state: "{{ host_collection_state }}"
+  register: result
+- fail:
+    msg: "Ensuring host collection is {{ host_collection_state }} failed! (expected_change: {{ expected_change | default('unknown') }})"
+  when: (expected_change is defined) and (result.changed != expected_change)
+...
```
Comment (mdellweg) [inline test/test_playbooks/tasks/host_collection.yml]: Newline?

File: test/test_playbooks/tasks/activation_key_copy.yml (hunk 12 above)
Comment (ephracis) [reply test/test_playbooks/tasks/activation_key_copy.yml]: Removing.

File: .gitignore (hunk 11 above)
Comment (ephracis) [reply .gitignore]: Sure thing.

File: test/test_playbooks/tasks/activation_key.yml (hunk 10 above)
Comment (ephracis) [reply test/test_playbooks/tasks/activation_key.yml]: Oops. 🤦‍♂

File: test/test_playbooks/tasks/host_collection.yml (hunk 13 above)
Comment (ephracis) [reply test/test_playbooks/tasks/host_collection.yml]: Adding.

**Comment by wzzrd** [comment]
> Nit picking phase. ;)
> Given the four last comments, i can say great work of both of you.
I don't think I had anything to do with this? 

> @wzzrd maybe you want to have a look, too.
Happy this is moving forward though. I can't wait for the day when this battlestation^Hthese modules will be fully functional.

**Pull request merged**